File: code/202-Happy Number/202-Happy Number.cpp
/*Happy Number
Write an algorithm to determine if a number is "happy".

A happy number is a number defined by the following process: Starting with any positive integer, replace the number by the sum of the squares of its digits, and repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle which does not include 1. Those numbers for which this process ends in 1 are happy numbers.

Example: 19 is a happy number

1^2 + 9^2 = 82
8^2 + 2^2 = 68
6^2 + 8^2 = 100
1^2 + 0^2 + 0^2 = 1

用set模拟，中间值都存到set里，如果循环了就false如果出1了就true

*/

class Solution {
public:
    int cal(int num){
        int res = 0;
        while (num){
            res += (num % 10) * (num % 10);
            num /= 10; 
        }
        return res;
    }

    bool isHappy(int n) {
       set<int> tmp;
    
        while (n != 1){
            if (tmp.find(n) == tmp.end())
                tmp.insert(n);
            else
                return false;
            n = cal(n);
        }
        return true;
    }
};

File: code/202-Happy Number/202-Happy Number.py
"""
LeetCode 202 - Happy Number

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
"""
# Write an algorithm to determine if a number n is "happy".
#
# A happy number is a number defined by the following process: Starting with any positive integer, replace the number by the sum of the squares of its digits, and repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle which does not include 1. Those numbers for which this process ends in 1 are happy numbers.
#
# Return True if n is a happy number, and False if not.
#
# Example:
#
# Input: 19
# Output: true
# Explanation:
# 12 + 92 = 82
# 82 + 22 = 68
# 62 + 82 = 100
# 12 + 02 + 02 = 1

class Solution(object):
    def cal(self, n):
        res = 0
        while n > 0:
            res += (n % 10) ** 2
            n = n / 10
        return res

    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        slow = fast = n
        while True:
            slow = self.cal(slow)
            fast = self.cal(fast)
            fast = self.cal(fast)
            if slow == fast:
                break
        return slow == 1

